pkg/apis/mysql/v1alpha1/types.go

```diff
@@ -0,0 +1,400 @@
+	// and server key for group replication SSL.
+	// +optional
+	SSLSecret *corev1.LocalObjectReference `json:"sslSecret,omitempty"`
+}
+
+// ClusterConditionType represents a valid condition of a Cluster.
+type ClusterConditionType string
+
+const (
+	// ClusterReady means the Cluster is able to service requests.
+	ClusterReady ClusterConditionType = "Ready"
+)
+
+// ClusterCondition describes the observed state of a Cluster at a certain point.
+type ClusterCondition struct {
+	Type   ClusterConditionType
+	Status corev1.ConditionStatus
+	// +optional
+	LastTransitionTime metav1.Time
+	// +optional
+	Reason string
+	// +optional
+	Message string
+}
+
+// ClusterStatus defines the current status of a MySQL cluster
+// propagating useful information back to the cluster admin
+type ClusterStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +optional
+	Conditions []ClusterCondition
+}
+
+// +genclient
+// +genclient:noStatus
+// +resourceName=mysqlclusters
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Cluster represents a cluster spec and associated metadata
+type Cluster struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              ClusterSpec   `json:"spec"`
+	Status            ClusterStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterList is a placeholder type for a list of MySQL clusters
+type ClusterList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []Cluster `json:"items"`
+}
+
+// BackupSpec defines the specification for a MySQL backup. This includes what should be backed up,
+// what tool should perform the backup, and, where the backup should be stored.
+type BackupSpec struct {
+	// Executor is the configuration of the tool that will produce the backup, and a definition of
+	// what databases and tables to backup.
+	Executor *BackupExecutor `json:"executor"`
+
+	// StorageProvider is the configuration of where and how backups should be stored.
+	StorageProvider *BackupStorageProvider `json:"storageProvider"`
+
+	// Cluster is a reference to the Cluster to which the Backup belongs.
+	Cluster *corev1.LocalObjectReference `json:"cluster"`
+
+	// AgentScheduled is the agent hostname to run the backup on.
+	// TODO(apryde): ScheduledAgent (*corev1.LocalObjectReference)?
+	AgentScheduled string `json:"agentscheduled"`
+}
+
+// BackupExecutor represents the configuration of the tool performing the backup. This includes the tool
+// to use, and, what database and tables should be backed up.
+// The storage of the backup is configured in the relevant Storage configuration.
+type BackupExecutor struct {
+	// Name of the tool performing the backup, e.g. mysqldump.
+	Name string `json:"name"`
+	// Databases are the databases to backup.
+	Databases []string `json:"databases"`
```

Comment:
Do you ever see this needing more fields under database? Like the table? If so i'd recommend using a struct with a name field.